packages/next/server/edge-functions/response.ts

```diff
@@ -0,0 +1,264 @@
+import { encode, byteLength } from './utils'
+import cookie, { CookieSerializeOptions } from 'next/dist/compiled/cookie'
+import type { Dictionary, HeadersEvent } from './types'
+
+interface Options {
+  method: string
+  headers?: Headers
+  onHeadersSent(event: HeadersEvent, res: EdgeResponse): void
+}
+
+export class EdgeResponse {
+  private _body?: null | string | ReadableStream
+  private _input = new TransformStream()
+  private _method: string
+  private _cookieParser: () => { [key: string]: string }
+  private _onHeadersSent: (event: HeadersEvent, res: EdgeResponse) => void
+  private _output = new TransformStream()
+  private _reader?: ReadableStreamDefaultReader
+  private _writer?: WritableStreamDefaultWriter
+  private _streaming = false
+
+  public finished = false
+  public headers
+  public headersSent = false
+  public statusCode = 200
+
+  constructor(options: Options) {
+    this._method = options.method
+    this.headers = options.headers || new Headers()
+    this._onHeadersSent = options.onHeadersSent
+
+    this._cookieParser = () => {
+      const value = this.headers.get('cookie')
+      return value ? cookie.parse(value) : {}
+    }
+
+    this.reader
+      .read()
+      .then(async (result) => {
+        this._streaming = true
+
+        if (!this.headersSent) {
+          this.writeHead(this.statusCode)
+        }
+
+        if (!result.done && result.value) {
+          await this.writer.write(result.value)
+          while (true) {
+            let { value, done } = await this.reader.read()
+            if (done) break
+            await this.writer.write(value)
+          }
+        }
+
+        this.writer.close()
+      })
+      .catch(console.error)
+  }
+
+  private get reader() {
+    if (!this._reader) {
+      this._reader = this._input.readable.getReader()
+    }
+
+    return this._reader
+  }
+
+  private get writer() {
+    if (!this._writer) {
+      this._writer = this._output.writable.getWriter()
+      this._streaming = true
+    }
+
+    return this._writer
+  }
+
+  private location(url: string) {
+    this.headers.set(
+      'Location',
+      url === 'back' ? this.headers.get('Referrer') || '/' : url
+    )
+  }
+
+  private _sendHeaders(event: HeadersEvent) {
+    if (this.headersSent) {
```

Comment:
@javivelasco Made this update here to prevent setting headers after headers are sent.